File: test/Models/pheno_pkg/src/openalea/Updatecalendar.py
# coding: utf8
from copy import copy

import numpy
from math import *
from datetime import datetime

def model_updatecalendar(cumulTT = 741.510096672,
         calendarMoments = ["Sowing"],
         calendarDates = [datetime(2007, 3, 21)],
         calendarCumuls = [0.0],
         currentdate = datetime(2007, 3, 27),
         phase = 1.0):
    """
     - Name: UpdateCalendar -Version: 1.0, -Time step: 1
     - Description:
                 * Title: Calendar Model
                 * Author: Pierre Martre
                 * Reference: Modeling development phase in the 
                     Wheat Simulation Model SiriusQuality.
                     See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
                 * Institution: INRA Montpellier
                 * Abstract: Lists containing for each stage the date it occurs as well as a copy of all types of cumulated thermal times 
     - inputs:
                 * name: cumulTT
                               ** min : -200
                               ** default : 741.510096671757
                               ** max : 10000
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : °C d
                               ** description : cumul thermal times at current date
                 * name: calendarMoments
                               ** variablecategory : state
                               ** datatype : STRINGLIST
                               ** default : ['Sowing']
                               ** inputtype : variable
                               ** unit : 
                               ** description : List containing apparition of each stage
                 * name: calendarDates
                               ** variablecategory : state
                               ** datatype : DATELIST
                               ** default : ['2007/3/21']
                               ** inputtype : variable
                               ** unit : 
                               ** description : List containing  the dates of the wheat developmental phases
                 * name: calendarCumuls
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** default : [0.0]
                               ** inputtype : variable
                               ** unit : °C d
                               ** description : list containing for each stage occured its cumulated thermal times
                 * name: currentdate
                               ** variablecategory : auxiliary
                               ** datatype : DATE
                               ** default : 2007/3/27
                               ** inputtype : variable
                               ** unit : 
                               ** description : current date
                 * name: phase
                               ** min : 0
                               ** default : 1
                               ** max : 7
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : 
                               ** description :  the name of the phase
     - outputs:
                 * name: calendarMoments
                               ** variablecategory : state
                               ** datatype : STRINGLIST
                               ** unit : 
                               ** description :  List containing apparition of each stage
                 * name: calendarDates
                               ** variablecategory : state
                               ** datatype : DATELIST
                               ** unit : 
                               ** description :  List containing  the dates of the wheat developmental phases
                 * name: calendarCumuls
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** unit : °C d
                               ** description :  list containing for each stage occured its cumulated thermal times
    """

    if phase >= 1.0 and phase < 2.0 and "Emergence" not in calendarMoments:
        calendarMoments.append("Emergence")
        calendarCumuls.append(cumulTT)
        calendarDates.append(currentdate)
    elif phase >= 2.0 and phase < 3.0 and "FloralInitiation" not in calendarMoments:
        calendarMoments.append("FloralInitiation")
        calendarCumuls.append(cumulTT)
        calendarDates.append(currentdate)
    elif phase >= 3.0 and phase < 4.0 and "Heading" not in calendarMoments:
        calendarMoments.append("Heading")
        calendarCumuls.append(cumulTT)
        calendarDates.append(currentdate)
    elif phase == 4.0 and "Anthesis" not in calendarMoments:
        calendarMoments.append("Anthesis")
        calendarCumuls.append(cumulTT)
        calendarDates.append(currentdate)
    elif phase == 4.5 and "EndCellDivision" not in calendarMoments:
        calendarMoments.append("EndCellDivision")
        calendarCumuls.append(cumulTT)
        calendarDates.append(currentdate)
    elif phase >= 5.0 and phase < 6.0 and "EndGrainFilling" not in calendarMoments:
        calendarMoments.append("EndGrainFilling")
        calendarCumuls.append(cumulTT)
        calendarDates.append(currentdate)
    elif phase >= 6.0 and phase < 7.0 and "Maturity" not in calendarMoments:
        calendarMoments.append("Maturity")
        calendarCumuls.append(cumulTT)
        calendarDates.append(currentdate)
    return (calendarMoments, calendarDates, calendarCumuls)

File: test/Models/pheno_pkg/src/f90/Updatecalendar.f90
MODULE Updatecalendarmod
    USE list_sub
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_updatecalendar(cumulTT, &
        calendarMoments, &
        calendarDates, &
        calendarCumuls, &
        currentdate, &
        phase)
        IMPLICIT NONE
        REAL, INTENT(IN) :: cumulTT
        CHARACTER(65), ALLOCATABLE , DIMENSION(:), INTENT(INOUT) ::  &
                calendarMoments
        CHARACTER(65), ALLOCATABLE , DIMENSION(:), INTENT(INOUT) ::  &
                calendarDates
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(INOUT) :: calendarCumuls
        CHARACTER(65), INTENT(IN) :: currentdate
        REAL, INTENT(IN) :: phase
        !- Description:
    !            * Title: Calendar Model
    !            * Author: Pierre Martre
    !            * Reference: Modeling development phase in the 
    !                Wheat Simulation Model SiriusQuality.
    !                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    !            * Institution: INRA Montpellier
    !            * Abstract: Lists containing for each stage the date it occurs as well as a copy of all types of cumulated thermal times 
        !- inputs:
    !            * name: cumulTT
    !                          ** description : cumul thermal times at current date
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** min : -200
    !                          ** max : 10000
    !                          ** default : 741.510096671757
    !                          ** unit : °C d
    !                          ** inputtype : variable
    !            * name: calendarMoments
    !                          ** description : List containing apparition of each stage
    !                          ** variablecategory : state
    !                          ** datatype : STRINGLIST
    !                          ** default : ['Sowing']
    !                          ** unit : 
    !                          ** inputtype : variable
    !            * name: calendarDates
    !                          ** description : List containing  the dates of the wheat developmental phases
    !                          ** variablecategory : state
    !                          ** datatype : DATELIST
    !                          ** default : ['2007/3/21']
    !                          ** unit : 
    !                          ** inputtype : variable
    !            * name: calendarCumuls
    !                          ** description : list containing for each stage occured its cumulated thermal times
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** default : [0.0]
    !                          ** unit : °C d
    !                          ** inputtype : variable
    !            * name: currentdate
    !                          ** description : current date
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DATE
    !                          ** default : 2007/3/27
    !                          ** unit : 
    !                          ** inputtype : variable
    !            * name: phase
    !                          ** description :  the name of the phase
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** min : 0
    !                          ** max : 7
    !                          ** default : 1
    !                          ** unit : 
    !                          ** inputtype : variable
        !- outputs:
    !            * name: calendarMoments
    !                          ** description :  List containing apparition of each stage
    !                          ** variablecategory : state
    !                          ** datatype : STRINGLIST
    !                          ** unit : 
    !            * name: calendarDates
    !                          ** description :  List containing  the dates of the wheat developmental phases
    !                          ** variablecategory : state
    !                          ** datatype : DATELIST
    !                          ** unit : 
    !            * name: calendarCumuls
    !                          ** description :  list containing for each stage occured its cumulated thermal times
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** unit : °C d
        IF(phase .GE. 1.0 .AND. phase .LT. 2.0 .AND. ALL(calendarMoments .NE.  &
                'Emergence')) THEN
            call Add(calendarMoments, 'Emergence')
            call Add(calendarCumuls, cumulTT)
            call Add(calendarDates, currentdate)
        ELSE IF ( phase .GE. 2.0 .AND. phase .LT. 3.0 .AND.  &
                ALL(calendarMoments .NE. 'FloralInitiation')) THEN
            call Add(calendarMoments, 'FloralInitiation')
            call Add(calendarCumuls, cumulTT)
            call Add(calendarDates, currentdate)
        ELSE IF ( phase .GE. 3.0 .AND. phase .LT. 4.0 .AND.  &
                ALL(calendarMoments .NE. 'Heading')) THEN
            call Add(calendarMoments, 'Heading')
            call Add(calendarCumuls, cumulTT)
            call Add(calendarDates, currentdate)
        ELSE IF ( phase .EQ. 4.0 .AND. ALL(calendarMoments .NE. 'Anthesis'))  &
                THEN
            call Add(calendarMoments, 'Anthesis')
            call Add(calendarCumuls, cumulTT)
            call Add(calendarDates, currentdate)
        ELSE IF ( phase .EQ. 4.5 .AND. ALL(calendarMoments .NE.  &
                'EndCellDivision')) THEN
            call Add(calendarMoments, 'EndCellDivision')
            call Add(calendarCumuls, cumulTT)
            call Add(calendarDates, currentdate)
        ELSE IF ( phase .GE. 5.0 .AND. phase .LT. 6.0 .AND.  &
                ALL(calendarMoments .NE. 'EndGrainFilling')) THEN
            call Add(calendarMoments, 'EndGrainFilling')
            call Add(calendarCumuls, cumulTT)
            call Add(calendarDates, currentdate)
        ELSE IF ( phase .GE. 6.0 .AND. phase .LT. 7.0 .AND.  &
                ALL(calendarMoments .NE. 'Maturity')) THEN
            call Add(calendarMoments, 'Maturity')
            call Add(calendarCumuls, cumulTT)
            call Add(calendarDates, currentdate)
        END IF
    END SUBROUTINE model_updatecalendar

END MODULE
